/*
 *  This is free software: you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *  The software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  You should have received a copy of the GNU General Public License
 *  along with the software. If not, see <http://www.gnu.org/licenses/>.
 *
 *  Copyright 2019, ACED.
 */


/**
 * @file This file contains the protobuf Epi output format.
 */

/**
 * Set version number.
 */
syntax = "proto3";

/**
 * Set package name. This will make sure that all the code
 * will be in the C++ namespace "proto".
 */
package proto;

/**
 * Encodes the entire Epidata file.
 */
message EpiFile {
    repeated Timestep timesteps = 1;
}

/**
 * Encodes a timestep. This consists of multiple Locations.
 */
message Timestep {
    repeated Location locations = 1;
}

/**
 * Encodes a location. This contains the name of the
 * location, the geographical coordinate of the location
 * and information about the different population sections
 * inside the location.
 */
message Location {
    // location data    
    string name      = 1;
    Coordinate coord = 2;

    // population data
    PopSection household = 3;
    PopSection k12school = 4;
    PopSection college   = 5;
    PopSection workplace = 6;
    PopSection primCom   = 7;
    PopSection secCom    = 8;
    PopSection daycare   = 9;
    PopSection preschool = 10;
}

/**
 * Encodes a geographical coordinate in the (long, lat) format.
 */
message Coordinate {
    double longitude = 1;
    double latitude = 2;
}

/**
 * Encodes a Population Section. This keeps track of the
 * number of individuals, and the fractions of the population
 * that correspond to the health statuses.
 */
message PopSection {
    uint32 population  = 1;
    double immune      = 2;
    double infected    = 3;
    double infectious  = 4;
    double recovered   = 5;
    double susceptible = 6;
    double symptomatic = 7;
}




